pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,464 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	feddnsv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/multiclusterdns/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+)
+
+const (
+	// minDNSTTL is the minimum safe DNS TTL value to use (in seconds).  We use this as the TTL for all DNS records.
+	minDNSTTL = 180
+
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	maxRetries    = 5
+
+	userAgent = "ServiceDNSEndpoint"
+)
+
+type ServiceDNSEndpointController struct {
+	// Client to federation api server
+	client fedclientset.Interface
+	// Informer Store for ServiceDNS objects
+	serviceDNSObjectStore cache.Store
+	// Informer controller for ServiceDNS objects
+	serviceDNSObjectController cache.Controller
+
+	queue         workqueue.RateLimitingInterface
+	minRetryDelay time.Duration
+	maxRetryDelay time.Duration
+}
+
+func StartController(config *restclient.Config, stopChan <-chan struct{}, minimizeLatency bool) error {
+	restclient.AddUserAgent(config, userAgent)
+	fedClient := fedclientset.NewForConfigOrDie(config)
+
+	controller, err := newServiceDNSEndpointController(fedClient, minimizeLatency)
+	if err != nil {
+		return err
+	}
+	go controller.Run(stopChan, 2)
+	return nil
+}
+
+func newServiceDNSEndpointController(client fedclientset.Interface, minimizeLatency bool) (*ServiceDNSEndpointController, error) {
+	d := &ServiceDNSEndpointController{
+		client: client,
+	}
+
+	// Start informer in federated API servers on DNS objects
+	// TODO: Change this to shared informer
+	d.serviceDNSObjectStore, d.serviceDNSObjectController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (pkgruntime.Object, error) {
+				return client.MulticlusterdnsV1alpha1().MultiClusterServiceDNSRecords(metav1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return client.MulticlusterdnsV1alpha1().MultiClusterServiceDNSRecords(metav1.NamespaceAll).Watch(options)
+			},
+		},
+		&feddnsv1a1.MultiClusterServiceDNSRecord{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: d.enqueueObject,
+			UpdateFunc: func(old, cur interface{}) {
+				oldObj, ok1 := old.(*feddnsv1a1.MultiClusterServiceDNSRecord)
+				curObj, ok2 := cur.(*feddnsv1a1.MultiClusterServiceDNSRecord)
+				if ok1 && ok2 && d.needsUpdate(oldObj, curObj) {
+					d.enqueueObject(cur)
+				}
+			},
+			DeleteFunc: d.enqueueObject,
+		},
+	)
+
+	d.minRetryDelay = minRetryDelay
+	d.maxRetryDelay = maxRetryDelay
+	if minimizeLatency {
+		d.minimizeLatency()
+	}
+	d.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(d.minRetryDelay, d.maxRetryDelay), userAgent)
+
+	netWrapper = &NetWrapperDefaultImplementation{}
```

Comment:
1. The declaration of variable netWrapper, should ideally be before this statement; it took me couple of minutes to search where this variable appeared here from.
2. It seems like u just want it initialized once, would `init()` be a more suitable location?

Reply:
Agree with your comments. updated accordingly. PTAL.